main.php

```diff
@@ -1,226 +1,19 @@
-			return;
-		}
-
-		// Fail and redirect the user to the login page
-		$loginPage = Director::absoluteURL(Security::config()->login_url);
-		$loginPage .= "?BackURL=" . urlencode($_SERVER['REQUEST_URI']);
-		header('location: '.$loginPage, true, 302);
-		die;
-	})
-	// Finally if a token was requested but there was an error while figuring out if it's allowed, do it anyway
-	->thenIfErrored(function() use ($reloadToken){
-		if ($reloadToken) {
-			$reloadToken->reloadWithToken();
-		}
-	})
-	->execute();
-
-global $databaseConfig;
-
-// Redirect to the installer if no database is selected
-if(!isset($databaseConfig) || !isset($databaseConfig['database']) || !$databaseConfig['database']) {
-
-    // Is there an _ss_environment.php file?
-    if(file_exists(BASE_PATH . '/_ss_environment.php') || file_exists(dirname(BASE_PATH) . '/_ss_environment.php')) {
-        header($_SERVER['SERVER_PROTOCOL'] . " 500 Server Error");
-        $dv = new SilverStripe\Dev\DebugView();
-        echo $dv->renderHeader();
-        echo $dv->renderInfo(
-            "Configuraton Error",
-            Director::absoluteBaseURL()
-        );
-        echo $dv->renderParagraph(
-            'You need to replace your _ss_environment.php file with a .env file, or with environment variables.<br><br>'
-            . 'See the <a href="https://docs.silverstripe.org/en/4/getting_started/environment_management/">'
-            . 'Environment Management</a> docs for more information.'
-        );
-        echo $dv->renderFooter();
-
-        die();
-    }
-
-	if(!file_exists(BASE_PATH . '/install.php')) {
-		header($_SERVER['SERVER_PROTOCOL'] . " 500 Server Error");
-		die('SilverStripe Framework requires a $databaseConfig defined.');
-	}
-	$host = isset($_SERVER['HTTP_HOST']) ? $_SERVER['HTTP_HOST'] : $_SERVER['SERVER_NAME'];
-	$s = (isset($_SERVER['SSL']) || (isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] != 'off')) ? 's' : '';
-	$installURL = "http$s://" . $host . BASE_URL . '/install.php';
-
-	// The above dirname() will equate to "\" on Windows when installing directly from http://localhost (not using
-	// a sub-directory), this really messes things up in some browsers. Let's get rid of the backslashes
-	$installURL = str_replace('\\', '', $installURL);
-
-	header("Location: $installURL");
-	die();
-}
-
-// Direct away - this is the "main" function, that hands control to the appropriate controller
-DataModel::set_inst(new DataModel());
-Director::direct($url, DataModel::inst());
+use SilverStripe\Core\AppKernel;
+use SilverStripe\Core\HTTPApplication;
+use SilverStripe\Core\Startup\ErrorControlChainMiddleware;
+use SilverStripe\Core\Startup\OutputMiddleware;
+use SilverStripe\Control\HTTPRequest;
+
+require __DIR__ . '/src/includes/autoload.php';
+
+// Build request and detect flush
+$request = HTTPRequest::createFromEnvironment();
+
+// Default application
+$kernel = new AppKernel();
+$app = new HTTPApplication($kernel);
+$app->addMiddleware(new OutputMiddleware());
```

Comment:
Here's what I've written on Slack before. Sorry to bring this up a third time Damian, discussions can easily trail off on Slack (less persistent than on a pull request).

---

Current approach

```$kernel = new AppKernel($flush);
$app = new HTTPApplication($kernel);
$app->addMiddleware(new OutputMiddleware());
$app->addMiddleware(new ErrorControlChainMiddleware($app, $request));
$app->handle($request);
```

How is this better than the following approach which is modeled after Symfony? https://github.com/symfony/symfony-demo/blob/master/web/app.php. Output isn’t a middleware in my opinion, it’s a function to be invoked on the resulting response. Or not, in case you’re doing testing. As far as I can tell, you can still catch exceptions in the same way?

```$kernel = new AppKernel($flush);
$app = new HTTPApplication($kernel);
$app->addMiddleware(new ErrorControlChainMiddleware($app, $request));
$response = $app->handle($request);
$response->output();
```

I’ve seen your `defaultResponse` logic in https://github.com/open-sausages/silverstripe-framework/blob/af601ca82c57d6c4d1c68a2feda05e5abf1d4a02/src/Core/Startup/OutputMiddleware.php. Symfony has the exception handling and fallback responses in `handle()`: https://github.com/symfony/symfony/blob/3.4/src/Symfony/Component/HttpKernel/HttpKernel.php#L63

All in all, I think Symfony’s HTTPFoundation stuff is solid. It’s used in other frameworks like Laravel. For various reasons we can’t just use it directly. But I think we should stick to that model as much as we can. There’s been a lot of thought into the Symfony code. We shouldn’t try to invent something different like `OutputMiddleware` unless there’s a really good reason for it.

For example, in `HttpKernel->handle()` you can opt out of the built-in exception handling via a method arg, and override `finishRequest()` in your own `AppKernel`. These are all things we’ll eventually have to do as well, so we might as well take inspiration from a project that has gone down that road already

---

Damian responded on Slack:

> I think it fits as a middleware. Moving it out, again, means you can't have "after response" middleware around request outputting. You can still easily do exactly what you proposed without any API changes (just modify index.php) so I don't see this approach as a limitation.

What's an example of this? If you want pseudo-async jobs, `register_shutdown_handler()` after output is a separate concern. And in terms of a *middle*ware for HTTP request/response, they shouldn't be involved after the output has taken place. For all you know, the response has been sent (ignoring output buffering), you shouldn't mess with it any more. In general, I see all of this being customised in a project specific `AppKernel`, rather than relying on declarative stuff like `$defaultResponse` arguments in a `OutputMiddleware`